Overhaul GDB Tutorial (#121)
eecs280staff/tutorials: docs/setup_gdb.md

GDB uses commands.  Here's a handy [reference card](http://users.ece.utexas.edu/~adnan/gdb-refcard.pdf).

## Quick Start
Compile your program with the `-g` flag and start GDB in text user interface (TUI) mode.
```console
$ make main.exe
g++ -Wall -Werror -pedantic -g --std=c++11 main.cpp -o main.exe
$ gdb -tui main.exe
```

| `b main` | breakpoint on main function |
| `b 13` | breakpoint on line 13 of current file |
| `b main.cpp:13` | breakpoint on line 13 of main.cpp |
| `r` | run or rerun |
| `l` | list a few lines of code |
| `n` | step over (AKA next) |
| `s` | step into |
| `up` | step out |
| `c` | continue |
| `p myvar` | print variable `myvar` |
| `refresh` | refresh TUI view |
| `bt` | backtrace, useful for segfaults |
| `q` | quit |

## Install
You should already have `g++` and `gdb` installed from the [WSL tutorial](setup_wsl.html).  Your versions might be different.
```console
$ g++ --version
g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
$ gdb --version
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
```

## Compile and Run
GDB uses an executable you build at the command line.

First, compile and run your executable at the command line.
```console

> $ make main.exe

Reviewer: ```suggestion
$ make main.exe
g++ -Wall -Werror -pedantic -g --std=c++11 main.cpp -o main.exe
```
I know this command is already shown in the Quick Start snippet above, but I think it's just nice to have the command there for people following along.

First, edit your `Makefile` and add the `CXXFLAGS` recommended by the [ASAN Quick Start](setup_asan.html#quick-start).

### Input redirection
If you're unfamiliar with input redirection, first read the CLI tutorial section on [input redirection](cli.html#input-redirection-).

Run with input redirection.  Make sure to add the name of your input file (`main_test.in` in this example).
```
$ gdb main.exe
...
(gdb) r < main_test.in
...
```

### Arguments and options
Arguments and options are inputs to a program typed at the command line.  Here's an example from EECS 280 Project 5:
```console
$ ./main.exe train_small.csv test_small.csv --debug
```
{: data-variant="no-line-numbers" data-highlight="1" }

- `main.exe` is the name of the program
- `train_small.csv` and `test_small.csv` are arguments
- `--debug` is an option

To run a program with options or arguments in LLDB, inc
```console
$ gdb main.exe
(gdb) r train_small.csv test_small.csv --debug
```
{: data-highlight="2" }

## Debug
In this section, we'll set a breakpoint, which pauses the debugger.  Then, we'll cover some of the options to continue execution.

> `n` **Next AKA Step Over**

Reviewer: ```suggestion
`n` **Next / Step Over:**
```
Unsure about AKA here and everywhere else. Thoughts on "Next / Step Over"? Same for all the other AKAs.
Also added colon for visual separateness.

Run one line of code, stepping _over_ any function calls by running the whole function in one step.

> `s` **Step AKA Step Into**

Reviewer: Refer to other AKA comment

Run one line of code, stepping _into_ any function calls to execute them line-by-line.

> `up` **Up AKA Step Out**

Reviewer: Refer to other AKA comment

Run the program until it returns from the current function (or until the next breakpoint).

> `c` **Continue**

Reviewer: ```suggestion
`c` **Continue:**
```

Run the program until the next breakpoint.

> `q` **Quit**

Reviewer: ```suggestion
`q` **Quit:**
```

Quit GDB.

### Example code

To get started, copy this example `main.cpp` into your editor.
```c++
#include <iostream>
#include <vector>
using namespace std;

double sum (const vector<double> &data) {
  double total = 0;

>   for (size_t i=0; i<data.size(); ++i) {

Reviewer: ```suggestion
  for (size_t i = 0; i < data.size(); ++i) {
```
nitpicking...

   │12                                                                         │
   │13      int main() {                                                       │
B+ │14        vector<double> data;                                             │
   │15        data.push_back(10);                                              │
   │16        data.push_back(20);                                              │
   │17        data.push_back(30);                                              │
  >│18        cout << "sum(data) = " << sum(data) << endl;                     │
   │19      }                                                                  │
   │20                                                                         │
   │21                                                                         │
   │22                                                                         │
   └───────────────────────────────────────────────────────────────────────────┘
native process 95465 In: main                                L18   PC: 0x400cd2 

(gdb) n
(gdb) n
(gdb) n
(gdb) n
```
{: data-highlight="11,19-22" }

<div class="primer-spec-callout info" markdown="1">

> **Pro-tip:** Hit Return to repeat your previous command.

Reviewer: ```suggestion
**Pro-tip:** Hit <kbd>Enter</kbd> to repeat your previous command.
```
Guessing Enter is more familiar than Return for Windows users.
The \<kbd> styling is just a thought.

</div>

### Inspect
Print the value of a variable with `p`.  If you have trouble viewing the contents of a `vector`, see the [Pretty Printing STL Containers with `gdb`](#pretty-printing-stl-containers-with-gdb) section.

```
(gdb) p data

> $1 = std::vector of length 3, capacity 4 = {10,	20, 30}

Author: Good idea.  I added a link.

```
{: data-highlight="1" }

### Step into

> Enter `s` AKA "Step" AKA "Step Into".  The cursor enters the `sum()` function.

Reviewer: ```suggestion
Enter `s` (Step / Step Into).  The cursor enters the `sum()` function.
```
Refer to previous AKA comment.

```
   ┌──main.cpp─────────────────────────────────────────────────────────────────┐
   │1       #include <iostream>                                                │
   │2       #include <vector>                                                  │
   │3       using namespace std;                                               │
   │4                                                                          │
   │5       double sum (const vector<double> &data) {                          │
  >│6         double total = 0;                                                │

>    │7         for (size_t i=0; i<data.size(); ++i) {                           │

Reviewer: ```suggestion
   │7         for (size_t i = 0; i < data.size(); ++i) {                       │
```
Consistency with nitpick above...

   │8           total += data[i];                                              │
   │9         }                                                                │
   │10        return total;                                                    │
   │11      }                                                                  │
   │12                                                                         │
   │13      int main() {                                                       │
   └───────────────────────────────────────────────────────────────────────────┘
native process 95465 In: sum                                 L6    PC: 0x400c02 

(gdb) s
```
{: data-highlight="7,18" }

### Step out

> Enter `up` AKA "Step Out".  The `sum()` function completes, and the program pauses again.

Reviewer: ```suggestion
Enter `up` (Step Out).  The `sum()` function completes, and the program pauses again.
```

B+ │14        vector<double> data;                                             │
   │15        data.push_back(10);                                              │
   │16        data.push_back(20);                                              │
   │17        data.push_back(30);                                              │
  >│18        cout << "sum(data) = " << sum(data) << endl;                     │
   │19      }                                                                  │
   │20                                                                         │
   │21                                                                         │
   │22                                                                         │
   │23                                                                         │
   │24                                                                         │
   │25                                                                         │
   └───────────────────────────────────────────────────────────────────────────┘
native process 95465 In: main                                L18   PC: 0x400cf0 

(gdb) up
```
{: data-highlight="7,18" }

### Continue

> Enter `c` AKA "Continue" to run the program to the next breakpoint, or the end, whichever comes first.

Reviewer: ```suggestion
Enter `c` (Continue) to run the program to the next breakpoint, or the end, whichever comes first.
```

```
(gdb) c
Continuing.
sum(data) = 60
[Inferior 1 (process 95465) exited normally]
```
{: data-highlight="1" }

### Quit
Quit GDB.  **Pro-tip:** `Control-D` will quit.

```
(gdb) q
```
{: data-highlight="1" }

## Pro-tips
Use the up arrow key to cycle through previously used commands.

> Use TAB completion to automatically complete the name of a command or a variable.

Reviewer: ```suggestion
Use <kbd>TAB</kbd> completion to automatically complete the name of a command or a variable.
```

> Use Emacs keyboard shortcuts to enter and edit your current command.

Reviewer: ```suggestion
Use [Emacs](setup_emacs.html) keyboard shortcuts to enter and edit your current command.
```
Consistency with LLDb tutorial
